src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,287 @@
+        if (parent) {
+            root = Hash::combine(*parent, root);
+        } else {
+            root = Hash::combine(root, filler.next());
+        }
+
+        d++;
+    }
+
+    // We may not have parents for ancestor trees, so we fill
+    // the rest in here.
+    while (d < depth) {
+        root = Hash::combine(root, filler.next());
+        d++;
+    }
+
+    return root;
+}
+
+// This constructs an authentication path into the tree in the format that the circuit
+// wants. The caller provides `filler_hashes` to fill in the uncle subtrees.
+template<size_t Depth, typename Hash>
+MerklePath IncrementalMerkleTree<Depth, Hash>::path(std::deque<Hash> filler_hashes) const {
+    if (!left) {
+        throw std::runtime_error("can't create an authentication path for the beginning of the tree");
+    }
+
+    PathFiller<Hash> filler(filler_hashes);
+
+    std::vector<Hash> path;
+    std::vector<bool> index;
+
+    if (right) {
+        index.push_back(true);
+        path.push_back(*left);
+    } else {
+        index.push_back(false);
+        path.push_back(filler.next());
+    }
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
+        if (parent) {
+            index.push_back(true);
+            path.push_back(*parent);
+        } else {
+            index.push_back(false);
+            path.push_back(filler.next());
+        }
+
+        d++;
+    }
+
+    while (d < Depth) {
+        index.push_back(false);
+        path.push_back(filler.next());
+        d++;
+    }
+
+    std::vector<std::vector<bool>> merkle_path;
+    BOOST_FOREACH(Hash b, path)
+    {
+        std::vector<unsigned char> hashv(b.begin(), b.end());
+        std::vector<bool> tmp_b;
+
+        libzerocash::convertBytesVectorToVector(hashv, tmp_b);
+
+        merkle_path.push_back(tmp_b);
+    }
+
+    std::reverse(merkle_path.begin(), merkle_path.end());
+    std::reverse(index.begin(), index.end());
+
+    return MerklePath(merkle_path, index);
+}
+
+template<size_t Depth, typename Hash>
+std::deque<Hash> IncrementalWitness<Depth, Hash>::uncle_train() const {
+    std::deque<Hash> uncles(filled.begin(), filled.end());
+
+    if (cursor) {
+        uncles.push_back(cursor->root(cursor_depth));
+    }
+
+    return uncles;
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalWitness<Depth, Hash>::append(Hash obj) {
+    if (cursor) {
+        cursor->append(obj);
+
+        if (cursor->is_complete(cursor_depth)) {
+            filled.push_back(cursor->root(cursor_depth));
+            cursor = boost::none;
+        }
+    } else {
+        cursor_depth = tree.next_depth(filled.size());
+
+        if (cursor_depth >= Depth) {
+            throw std::runtime_error("tree is full");
+        }
+
+        if (cursor_depth == 0) {
+            filled.push_back(obj);
+        } else {
+            cursor = IncrementalMerkleTree<Depth, Hash>();
+            cursor->append(obj);
+        }
+    }
+}
```

Comment:
It's really hard for me to wrap my head around how `tree.next_depth()` and `uncle_train()` and `filler_hashes` interact. I wonder if there's a way to make the connection more obvious, but I don't want to block this PR on that.

Let's see, I find it easiest to think about it this way. We're witnessing an object at some position in the tree. Call the largest complete subtree containing obj at the time of witnessing T. T has a sibling T' (of the same depth as T) to its right which is empty at the time of witnessing. So we put new objects in there until it gets full. Once it's full, T and T' are both full so the parent of T and T' is full, too. We save the root of T'. Now we repeat the whole thing. We find the largest filled subtree containing the witnessed object (the parent of T and T'), call it U. U has a right sibling U' which is empty, so we add things there, until it gets full... repeat.

Okay, I understand it now.